crates/synthesis/src/two_qubit_decompose.rs

```diff
@@ -1947,27 +1913,127 @@ impl TwoQubitBasisDecomposer {
+        Ok(TwoQubitGateSequence {
+            gates,
+            global_phase,
+        })
+    }
+    /// Decompose a two-qubit ``unitary`` over fixed basis and :math:`SU(2)` using the best
+    /// approximation given that each basis application has a finite ``basis_fidelity``.
+    fn generate_sequence(
+        &self,
+        unitary: PyReadonlyArray2<Complex64>,
+        basis_fidelity: Option<f64>,
+        approximate: bool,
+        _num_basis_uses: Option<u8>,
+    ) -> PyResult<TwoQubitGateSequence> {
+        let basis_fidelity = if !approximate {
+            1.0
+        } else {
+            basis_fidelity.unwrap_or(self.basis_fidelity)
+        };
+        let target_decomposed =
+            TwoQubitWeylDecomposition::new(unitary, Some(DEFAULT_FIDELITY), None)?;
+        let traces = self.traces(&target_decomposed);
+        let best_nbasis = traces
+            .into_iter()
+            .enumerate()
+            .map(|(idx, trace)| (idx, trace.trace_to_fid() * basis_fidelity.powi(idx as i32)))
+            .min_by(|(_idx1, fid1), (_idx2, fid2)| fid2.partial_cmp(fid1).unwrap())
+            .unwrap()
+            .0;
+        let best_nbasis = _num_basis_uses.unwrap_or(best_nbasis as u8);
+        let decomposition = match best_nbasis {
+            0 => decomp0_inner(&target_decomposed),
+            1 => self.decomp1_inner(&target_decomposed),
+            2 => self.decomp2_supercontrolled_inner(&target_decomposed),
+            3 => self.decomp3_supercontrolled_inner(&target_decomposed),
+            _ => unreachable!("Invalid basis to use"),
+        };
+        let pulse_optimize = self.pulse_optimize.unwrap_or(true);
+        let sequence = if pulse_optimize {
+            self.pulse_optimal_chooser(best_nbasis, &decomposition, &target_decomposed)?
+        } else {
+            None
+        };
+        if let Some(seq) = sequence {
+            return Ok(seq);
+        }
+        let mut target_1q_basis_list = EulerBasisSet::new();
+        target_1q_basis_list.add_basis(self.euler_basis);
+        let euler_decompositions: SmallVec<[Option<OneQubitGateSequence>; 8]> = decomposition
+            .iter()
+            .map(|decomp| {
+                unitary_to_gate_sequence_inner(
+                    decomp.view(),
+                    &target_1q_basis_list,
+                    0,
+                    None,
+                    true,
+                    None,
+                )
+            })
+            .collect();
+        // Worst case length is 5x 1q gates for each 1q decomposition + 1x 2q gate
+        // We might overallocate a bit if the euler basis is different but
+        // the worst case is just 16 extra elements with just a String and 2 smallvecs
+        // each. This is only transient though as the circuit sequences aren't long lived
+        // and are just used to create a QuantumCircuit or DAGCircuit when we return to
+        // Python space.
+        let mut gates = Vec::with_capacity(21);
```

Comment:
Should we pull this out into a constant to make it more discoverable in the future?